### PR TITLE
fromFile function add

### PR DESCRIPTION
**Pull request by maddhruv**
Function to add entries from a JSON file

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -29,6 +29,17 @@ class Glossary {
     this._entries[term] = description
   }
 
+  fromFile (filename) {
+    let rawdata = fs.readFileSync(filename)
+    let glossary = JSON.parse(rawdata)
```
Comment (zeke) [inline index.js]: If we're only supporting JSON, this could just a be a `require` instead.

File: index.js [hunk 2]
```diff
@@ -29,6 +29,17 @@ class Glossary {
...
+    for (var i in glossary) {
+      assert(i && i.length, `term is required for description ${glossary[i]}`)
```
Comment (zeke) [inline index.js]: Rather than re-doing this validation, we should just use the existing `.add()` method and get the existing validation for free.

File: readme.md [hunk 3]
```diff
@@ -51,6 +51,13 @@ call `glossary.upload()`
 - `term` String (required)
 - `description` String (required)
 
+### `glossary.fromFile(filename)`
+
+Uploads entries from the a file provided. Entries only exist in memory until you
+call `glossary.upload()` but file remains unchanged.
```
Comment (zeke) [inline readme.md]: "but file remains unchanged." doesn't seem necessary.

File: readme.md [hunk 4]
```diff
@@ -51,6 +51,13 @@ call `glossary.upload()`
...
+
+- `filename` String (required)
```
Comment (zeke) [inline readme.md]: Might want to mention that this should be a full path.

**Comment by zeke** [comment]
Hey @maddhruv wanna take another look at this when you get a chance?

**Comment by maddhruv** [comment]
@zeke I would surely make relevant changes today

**Comment by maddhruv** [comment]
the build is failing as I have referred a non-existing dummy file

so should I also add the dummy file in the repo?

**Comment by zeke** [comment]
> should I also add the dummy file in the repo?

Yeah this is typically referred to as a "test fixture". It can be included in the repo itself in a file like `test/fixture.json` or you could create the file dynamically during the test and remove it after the test is finished.

**Comment by zeke** [comment]
That's a good start. You'll also want to test the "happy path" scenario, where the API works as expected.

Try using this command to help ensure you have good test coverage:

```
jest --watch --notify --notifyMode=change --coverage
```

(We should probably also add this command to the npm scripts for convenience)

**Comment by maddhruv** [comment]
I am poor at testing!
Haven't implemented them in codes before, honestly!
Will check that too

**Comment by maddhruv** [comment]
@zeke can this be merged now?

screenshot from 2018-06-01 09-17-31

**Comment by zeke** [comment]
> You'll also want to test the "happy path" scenario, where the API works as expected.

☝️ @maddhruv I'd love to see this implemented before we land it.

**Comment by maddhruv** [comment]
> You'll also want to test the "happy path" scenario, where the API works as expected.
I do not get ti :confused:

**Comment by zeke** [comment]
You have a test that verifies the behavior when the "wrong thing" happens (empty file), but no test that verifies when the "right thing" happens (non-empty file). This is sometimes called the "happy path", meaning everything works as expected. You should add a test that has a non-empty JSON file and verify that the `fromFile` function behaves as expected when loading it.

**Comment by maddhruv** [comment]
:smile_cat:

File: index.js [hunk 5]
```diff
@@ -29,6 +29,16 @@ class Glossary {
...
   }
 
+  fromFile (filename) {
+    let glossary = require(filename)
+    assert(Object.keys(glossary).length, 'The file seems to be empty')
+    glossary.forEach(term => {
+      assert(term, 'Term or description not found')
+      this.add(term[0], term[1])
+    })
+    this._entries = glossary
```
Comment (zeke) [inline index.js]: `this._entries = glossary` is not necessary. This line can be removed.

File: index.js [hunk 6]
```diff
@@ -29,6 +29,16 @@ class Glossary {
     this._entries[term] = description
   }
 
+  fromFile (filename) {
+    let glossary = require(filename)
+    assert(Object.keys(glossary).length, 'The file seems to be empty')
+    glossary.forEach(term => {
+      assert(term, 'Term or description not found')
```
Comment (zeke) [inline index.js]: `assert` is not necessary. This is already handled by the `add` method.

File: test/empty-file.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (zeke) [inline test/empty-file.json]: The docs and the implementation expect an array of arrays, not an object.

File: test/glossary_test.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (zeke) [inline test/glossary_test.json]: This should also be an array, not an object.

File: test.js [hunk 9]
```diff
@@ -64,6 +64,22 @@ describe('glossary.add()', () => {
...
+  const glossary = Glossary({project: 'bar', crowdinKey: 'xyz'})
+
+  test('is empty', () => {
+    expect(() => {
+      glossary.fromFile('./test/empty-file.json')
+    }).toThrow('The file seems to be empty')
+  })
+
+  test('upload from a file', () => {
+    expect(() => {
```
Comment (zeke) [inline test.js]: This should be something like:

```js
glossary.fromFile('./test/glossary_test.json')
expect(glossary.entries.length).toEq(2)
```
